### ISSUE TITLE
[Phase 5.4] Wave Function Collapse Generator

### ISSUE THREAD

**Issue by leifrogers** [leifrogers/sashiko-designer#38]
## Description
Study existing traditional patterns to extract tiles, learn compatibility rules, and generate novel coherent variations that maintain traditional symmetries. Novel but authentic-feeling results through constraint satisfaction.

## Research Basis
Wave Function Collapse algorithm analyzes example patterns to generate infinite variations that feel coherent with the source material.

## Features
- Analyze example patterns to extract tiles
- Learn compatibility rules between tiles
- Generate novel patterns maintaining traditional symmetries
- Tile-based generation system
- Coherent variation that feels authentic
- Parameters:
  - Tile size
  - Output dimensions
  - Constraint strictness
- Use traditional patterns as training data
- Generate infinite variations

## Implementation
```javascript
class WaveFunctionCollapse {
  constructor(examples, tileSize) {
    this.tileSize = tileSize;
    this.tiles = [];
    this.adjacencyRules = {};
    
    // Analyze examples to extract tiles and rules
    this.analyzeSamples(examples);
  }
  
  analyzeSamples(examples) {
    // Extract all unique tiles from example patterns
    for (let example of examples) {
      this.extractTiles(example);
    }
    
    // Learn which tiles can be adjacent
    this.learnAdjacencyRules(examples);
  }
  
  extractTiles(pattern) {
    let w = pattern.width;
    let h = pattern.height;
    
    for (let y = 0; y < h - this.tileSize; y++) {
      for (let x = 0; x < w - this.tileSize; x++) {
        let tile = this.getTile(pattern, x, y);
        let tileId = this.getTileSignature(tile);
        
        if (!this.tiles.find(t => t.id === tileId)) {
          this.tiles.push({
            id: tileId,
            data: tile,
            rotations: this.generateRotations(tile)
          });
        }
      }
    }
  }
  
  learnAdjacencyRules(examples) {
    // For each tile, learn which tiles can be adjacent in each direction
    for (let example of examples) {
      for (let y = 0; y < example.height - this.tileSize; y++) {
        for (let x = 0; x < example.width - this.tileSize; x++) {
          let tile = this.getTileSignature(this.getTile(example, x, y));
          
          // Check right neighbor
          if (x + this.tileSize < example.width - this.tileSize) {
            let right = this.getTileSignature(this.getTile(example, x + this.tileSize, y));
            this.addAdjacency(tile, right, 'right');
          }
          
          // Check bottom neighbor
          if (y + this.tileSize < example.height - this.tileSize) {
            let bottom = this.getTileSignature(this.getTile(example, x, y + this.tileSize));
            this.addAdjacency(tile, bottom, 'bottom');
          }
        }
      }
    }
  }
  
  addAdjacency(tileA, tileB, direction) {
    if (!this.adjacencyRules[tileA]) {
      this.adjacencyRules[tileA] = { right: [], bottom: [] };
    }
    if (!this.adjacencyRules[tileA][direction].includes(tileB)) {
      this.adjacencyRules[tileA][direction].push(tileB);
    }
  }
  
  generate(outputWidth, outputHeight, constraintStrictness = 1.0) {
    // Initialize output grid with all possible tiles
    let grid = [];
    for (let y = 0; y < outputHeight; y++) {
      grid[y] = [];
      for (let x = 0; x < outputWidth; x++) {
        grid[y][x] = {
          collapsed: false,
          possibilities: this.tiles.map(t => t.id)
        };
      }
    }
    
    // Wave Function Collapse algorithm
    while (!this.isFullyCollapsed(grid)) {
      // Find cell with minimum entropy (fewest possibilities)
      let cell = this.findMinEntropyCell(grid);
      if (!cell) break;
      
      // Collapse it to one possibility
      this.collapse(cell, constraintStrictness);
      
      // Propagate constraints to neighbors
      this.propagate(grid, cell.x, cell.y);
    }
    
    return this.renderGrid(grid);
  }
  
  findMinEntropyCell(grid) {
    let minEntropy = Infinity;
    let minCell = null;
    
    for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[y].length; x++) {
        let cell = grid[y][x];
        if (!cell.collapsed && cell.possibilities.length < minEntropy) {
          minEntropy = cell.possibilities.length;
          minCell = { x, y, cell };
        }
      }
    }
    
    return minCell;
  }
  
  collapse(cell, strictness) {
    // Choose one possibility based on weights
    let possibilities = cell.cell.possibilities;
    
    if (strictness < 1.0) {
      // Allow some random variation
      if (random() > strictness) {
        possibilities = this.tiles.map(t => t.id);
      }
    }
    
    let chosen = random(possibilities);
    cell.cell.possibilities = [chosen];
    cell.cell.collapsed = true;
  }
  
  propagate(grid, x, y) {
    // Update neighbors based on collapsed cell
    let stack = [[x, y]];
    
    while (stack.length > 0) {
      let [cx, cy] = stack.pop();
      let currentTile = grid[cy][cx].possibilities[0];
      
      // Check all neighbors
      let directions = [
        { dx: 1, dy: 0, dir: 'right' },
        { dx: -1, dy: 0, dir: 'left' },
        { dx: 0, dy: 1, dir: 'bottom' },
        { dx: 0, dy: -1, dir: 'top' }
      ];
      
      for (let { dx, dy, dir } of directions) {
        let nx = cx + dx;
        let ny = cy + dy;
        
        if (nx >= 0 && nx < grid[0].length && ny >= 0 && ny < grid.length) {
          let neighbor = grid[ny][nx];
          if (!neighbor.collapsed) {
            let validTiles = this.adjacencyRules[currentTile]?.[dir] || [];
            let newPossibilities = neighbor.possibilities.filter(t => validTiles.includes(t));
            
            if (newPossibilities.length < neighbor.possibilities.length) {
              neighbor.possibilities = newPossibilities;
              stack.push([nx, ny]);
            }
          }
        }
      }
    }
  }
  
  isFullyCollapsed(grid) {
    for (let row of grid) {
      for (let cell of row) {
        if (!cell.collapsed) return false;
      }
    }
    return true;
  }
  
  renderGrid(grid) {
    // Convert collapsed grid to stitches
    let pattern = [];
    
    for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[y].length; x++) {
        let tileId = grid[y][x].possibilities[0];
        let tile = this.tiles.find(t => t.id === tileId);
        pattern.push({
          x: x * this.tileSize,
          y: y * this.tileSize,
          tile: tile.data
        });
      }
    }
    
    return pattern;
  }
}
```

## UI Components
- WFC generator panel
- Training pattern selector (use existing patterns)
- Tile size slider (8-32px)
- Output dimensions inputs
- Constraint strictness slider (0-100%)
- Generate button
- Training set manager:
  - Add pattern to training set
  - Remove from training set
  - View training set
- Generation preview
- Analyze training button

## Acceptance Criteria
- [ ] WFC algorithm implemented
- [ ] Tile extraction from examples working
- [ ] Adjacency rule learning functional
- [ ] Pattern generation working
- [ ] Coherent variations produced
- [ ] Traditional symmetries maintained
- [ ] Configurable tile size (8-32px)
- [ ] Configurable output dimensions
- [ ] Constraint strictness control
- [ ] Training set management
- [ ] Multiple traditional patterns usable as training
- [ ] Generated patterns feel authentic
- [ ] Infinite variations possible
- [ ] Performance acceptable for reasonable sizes

## Dependencies
- Requires #22 (Pattern Generator Architecture)
- Requires #21, #23-26 (Traditional patterns as training data)

## Estimate
~20-30 hours

## Priority
**Nice-to-Have (v3.0+)** - Week 10

## Labels
`enhancement`, `phase-5`, `v3.0`, `advanced`, `generative`, `wfc`
